File: DataStructures/com/binaryTrees/TopView.java
package com.binaryTrees;

import java.util.HashMap;
import java.util.Map;

public class TopView {

    public void topView(TreeNode root, int dist, int level, Map<Integer, Pair<Integer, Integer>> map) {

        if( root == null ) {
            return;
        }

        if( !map.containsKey(dist) || level < map.get(dist).second) {
            map.put(dist, Pair.of(root.data, level));
        }

        topView(root.left, dist - 1, level + 1, map);

        topView(root.right, dist + 1, level + 1, map);
    }

    public void printTopView(TreeNode root) {
        Map<Integer, Pair<Integer, Integer>> map = new HashMap<>();
        topView(root, 0, 0, map);

        for( Map.Entry<Integer, Pair<Integer, Integer>> entry: map.entrySet() ) {
            System.out.print(entry.getValue().first + " ");
        }
        System.out.println();

        for(Pair<Integer, Integer> it: map.values()) {
            System.out.print(it.first + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        BinaryTree binaryTree = new BinaryTree();
        TreeNode x = binaryTree.getBinaryTree();
       TopView tp = new TopView();
       tp.printTopView(x);
    }
}

File: TS/BinaryTrees/src/DataStructures/BinaryTree/BinaryTreeTraversals/views/topView.ts
import { TreeNode } from "../../TreeNode";
import BinaryTree from "../../TreeNode";
import Pair from "./Pair";


class TopView {


  private topView( root: TreeNode | null, dist: number, level: number,  
    map: Map< number, Pair< number, number >> ): void {
       
     if( root === null ) {
      return;
     }

     if( !map.has(dist) || level < map.get( dist )!.second) {
       map.set( dist, Pair.of(root.data, level));
     }

     if( root.left !== null ) this.topView(root.left, dist - 1, level + 1, map);
     if( root.right !== null ) this.topView( root.right, dist + 1, level + 1, map);

     
  }
   
   getTopView( root: TreeNode | null ): Array<number> {
     
     const result: Array<number> = [];
     if( root === null) {
      return result;
     }

      const map: Map<number, Pair<number, number>> = new Map();
      this.topView( root, 0, 1, map);
       for( const value of map.values()) {
         result.push( value.first );
       }
      return result;
   }
}

const binaryTree: BinaryTree = new BinaryTree(); 
const root: TreeNode = binaryTree.getBinary();

const topView: TopView = new TopView();

console.log( topView.getTopView( root ));
